src/audio/smart_amp/smart_amp.c

```diff
@@ -646,25 +629,95 @@ static int smart_amp_copy(struct comp_dev *dev)
 static int smart_amp_reset(struct comp_dev *dev)
 {
 	struct smart_amp_data *sad = comp_get_drvdata(dev);
+	struct smart_amp_mod_data_base *mod = sad->mod_data;
+	int ret;
 
 	comp_dbg(dev, "smart_amp_reset()");
 
-	sad->process = NULL;
-	sad->in_channels = 0;
-	sad->out_channels = 0;
+	sad->ff_get_frame = NULL;
+	sad->fb_get_frame = NULL;
+	sad->ff_set_frame = NULL;
+
+	/* reset inner model */
+	ret = mod->mod_ops->reset(mod);
+	if (ret)
+		return ret;
 
 	comp_set_state(dev, COMP_TRIGGER_RESET);
 
 	return 0;
 }
 
+static int smart_amp_set_frame_fmts(struct comp_dev *dev, uint16_t ff_src_fmt, uint16_t fb_src_fmt)
+{
+	struct smart_amp_data *sad = comp_get_drvdata(dev);
+	struct smart_amp_mod_data_base *mod = sad->mod_data;
+	const uint16_t *ff_fmts, *fb_fmts;
+	int num_ff_fmts, num_fb_fmts;
+	int ret;
+	int i;
+
+	/* get supported formats from mod */
+	ret = mod->mod_ops->get_supported_fmts(mod, &ff_fmts, &num_ff_fmts,
+					       &fb_fmts, &num_fb_fmts);
+	if (ret) {
+		comp_err(dev, "smart_amp_set_frame_fmts(): failed to get supported formats");
+		return ret;
+	}
+
+	/* available formats: SOF_IPC_FRAME_S16_LE, SOF_IPC_FRAME_S24_4LE, SOF_IPC_FRAME_S32_LE */
+	for (i = 0; i < num_ff_fmts; i++) {
+		if (ff_fmts[i] >= ff_src_fmt && ff_fmts[i] <= SOF_IPC_FRAME_S32_LE) {
+			comp_dbg(dev, "smart_amp_set_frame_fmts(): feed forward fmt src: %u, mod: %u",
+				 ff_src_fmt, ff_fmts[i]);
+			sad->ff_mod.frame_fmt = ff_fmts[i];
+			sad->out_mod.frame_fmt = ff_fmts[i];
+			break;
+		}
+	}
+
+	if (i == num_ff_fmts) {
+		comp_err(dev, "smart_amp_set_frame_fmts(): no supported feed forward fmt");
+		return -EINVAL;
+	}
+
+	for (i = 0; i < num_fb_fmts; i++) {
+		if (fb_fmts[i] >= fb_src_fmt && fb_fmts[i] <= SOF_IPC_FRAME_S32_LE) {
+			comp_dbg(dev, "smart_amp_set_frame_fmts(): feedback fmt src: %u, mod: %u",
+				 fb_src_fmt, fb_fmts[i]);
+			sad->fb_mod.frame_fmt = fb_fmts[i];
+			break;
+		}
+	}
+	if (i == num_fb_fmts) {
+		comp_err(dev, "smart_amp_set_frame_fmts(): no supported feedback fmt");
+		return -EINVAL;
+	}
+
+	/* set frame formats to inner model */
+	ret = mod->mod_ops->set_fmts(mod, sad->ff_mod.frame_fmt, sad->fb_mod.frame_fmt);
```

Comment:
I'm guessing we are missing a ```module_set_fmt()``` wrapper fo this in module core ?

Reply:
I'm not sure I get the idea. Do you mean we need the ops `set_fmt()` to set the format of core?

The design now has ops `set_fmts()` to set the format of feed-forward (FF) data path and feedback (FB) path respectively. Indeed the core format can be independent of input/ouput data format, however I wonder if that has the real need (in fact I think FF and FB formats could align each other (as well as core format) then `set_fmts()` could be replaced with `set_fmt()`). Since the design now is proposed to handle input/output data format conversion in generic layer, it should be fair to require the format alignment.

Reply:
Sorry, I meant an inline static wrapper func around the API ops dereference.

Reply:
I see. Anyway it triggered my re-consideration of the format for module core. I revised to unify the module formats which include core, ff-input and fb-input previously. Now the module ops will only set one format .